kernel/kernel_drivers.go

```diff
@@ -179,34 +202,129 @@ func RemoveKernelDriversTree(ksnapName string, rev snap.Revision) (err error) {
 	return os.RemoveAll(treeRoot)
 }
 
-// EnsureKernelDriversTree creates a tree in
-// /var/lib/snapd/kernel/<ksnapName>/<rev> with subfolders can be bind-mounted
-// on boot to /usr/lib/{modules,firmware}. This tree contains files from the
-// kernel snap mounted on kernelMount, as well as symlinks to it.
-// TODO this will be extended to consider kernel-modules components.
-func EnsureKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string) (err error) {
+// KernelModulesMntInfo contains information about a component and where it is
+// mounted.
+type KernelModulesMntInfo struct {
+	CompSideInfo *snap.ComponentSideInfo
+	MntDir       string
+}
+
+type KernelDriversTreeOptions struct {
+	// Set if we are building the tree for a kernel we are installing right now
+	KernelInstall bool
+}
+
+// EnsureKernelDriversTree creates a drivers tree that can include modules/fw
+// from kernel-modules components. opts.KernelInstall tells the function if
+// this is a kernel install (which might be installing components at the same
+// time) or an only components install.
+//
+// For kernel installs, a tree in /var/lib/snapd/kernel/<ksnapName>/<rev> with
+// subfolders can be bind-mounted on boot to /usr/lib/{modules,firmware} is
+// created. This tree contains files from the kernel snap mounted on
+// kernelMount, as well as symlinks to it. It gets used after a reboot.
+//
+// For components-only install, we want the components to be available without
+// rebooting. For this, we work on a temporal tree, and after finishing it we
+// swap atomically the affected modules/firmware folders with those of the
+// currently active kernel.
+func EnsureKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string, kmodsInfos []*KernelModulesMntInfo, opts *KernelDriversTreeOptions) (err error) {
+	// The temporal dir when installing only components can be fixed as a
+	// task installing/updating a kernel-modules component must conflict
+	// with changes containing this same task. This helps with clean-ups if
+	// something goes wrong. Note that this folder needs to be in the same
+	// filesystem as the final one so we can atomically switch the folders.
+	ksnapDir := ksnapName + "_tmp"
+	if opts.KernelInstall {
+		ksnapDir = ksnapName
+	}
 	// Initial clean-up to make the function idempotent
-	if rmErr := RemoveKernelDriversTree(ksnapName, rev); rmErr != nil &&
+	if rmErr := RemoveKernelDriversTree(ksnapDir, rev); rmErr != nil &&
```

Comment:
RemoveKernelDriversTree seems to assume that the first argument is the snap name, but we're sometimes passing `<snap-name>_tmp`. Perhaps its argument names need to be tweaked now.

Reply:
Changed